src/event.rs

```diff
@@ -53,6 +54,29 @@ impl Event {
         unsafe { ffi::gdk_event_put(self.to_glib_none().0) }
     }
 
+    /// Set the event handler.
+    /// 
+    /// The callback `handler` is called for each event. If `None`, event
+    /// handling is disabled.
+    pub fn set_handler<F: Fn(&mut Event) + Send +'static>(handler: Option<F>) {
+        if let Some(handler) = handler {
+            // allocate and convert to target type
+            // double box to reduce a fat pointer to a simple pointer
+            let boxed: Box<Box<Fn(&mut Event) + Send + 'static>> = Box::new(Box::new(handler));
+            let ptr: *mut c_void = Box::into_raw(boxed) as *mut _;
+            unsafe{
+                ffi::gdk_event_handler_set(
+                    Some(event_handler_trampoline),
+                    ptr,
+                    Some(event_handler_destroy))
+            }
+        } else {
+            unsafe{
+                ffi::gdk_event_handler_set(None, ptr::null_mut(), None)
```

Comment:
How thread-safe is `gdk_event_handler_set()`? Can it be called from multiple threads at once safely, i.e. is it internally protected by a mutex? If not then this function still needs to be marked as `unsafe`.

Reply:
Here it would impossible to deadlock but that's not helping.

I assume it's only allowed to call from the main thread, so we could use the assertion here that is used all over GTK for making sure things are only called from the main thread. Unfortunately that part of the code is in the gtk crate, so not sure how to easily get that over here.